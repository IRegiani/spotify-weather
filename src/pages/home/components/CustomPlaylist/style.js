import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles({
  root: {
    minWidth: '250px',
    maxWidth: '650px',
    maxHeight: '350px',
    textAlign: 'start',
    margin: '20px',
    display: 'grid',
    justifyContent: 'start',
    overflowY: 'auto',
  },
  title: {
    fontSize: 18,
    textTransform: 'capitalize',

  },
  description: {
    textTransform: 'capitalize',
  },
  skeletonCircle: {
    margin: '30px auto',
    marginBottom: '35px',
  },
  skeletonText: {
    margin: '8px auto',
    height: '1.7rem',
  },
  inline: {
    display: 'inline',
  },
});
