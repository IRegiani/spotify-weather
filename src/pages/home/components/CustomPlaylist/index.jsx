import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Card, CardContent, List, ListItem, ListItemAvatar, ListItemText, Divider, Avatar, Button, InputBase, Tooltip } from '@material-ui/core';
import DeleteOutlinedIcon from '@material-ui/icons/DeleteOutlined';
import SaveIcon from '@material-ui/icons/Save';

import { createPlaylist } from '../../../../api/spotify';
import { useStyles } from './style';

// TODO: Add description

export const savePlaylist = async (accessToken, description, playlistName, setLoading, { id }, tracks, disableButton) => {
  await createPlaylist(accessToken, id, playlistName, description, tracks);
  setLoading(false);
  disableButton(true);
};

const generateListItem = (listSize, { removeTrack }) => (track, index) => {
  const secondaryComponent = `${track.album.name} | ${track.artists.map((artist, inx) => (track.artists.length - 1 === inx ? artist.name : `${artist.name} - `))}`;

  return (
    <div key={track.id}>
      <ListItem alignItems="flex-start">
        <ListItemAvatar>
          <Avatar src={track.album.images[0].url} alt="track-cover" />
        </ListItemAvatar>
        <ListItemText primary={track.name} secondary={secondaryComponent} />
        <DeleteOutlinedIcon onClick={() => removeTrack(track)} />
      </ListItem>
      {index === listSize - 1 ? null : <Divider />}
    </div>
  );
};

const CustomPlaylistsWidget = ({ playlist, updatePlaylist, userProfile, accessToken }) => {
  const [playlistName, setPlaylistName] = useState('My Playlist');
  const [isLoading, setLoading] = useState(false);
  const [isButtonDisabled, setButtonDisabled] = useState(false);
  const classes = useStyles();

  const onSaveClick = () => { setLoading(true); savePlaylist(accessToken, undefined, playlistName, setLoading, userProfile, playlist, setButtonDisabled); };

  return (
    <Card variant="outlined" className={classes.root}>
      <CardContent>
        <Tooltip title="Rename your playlist!" leaveDelay={500} placement="right" arrow>
          <InputBase
            className={classes.input}
            value={playlistName}
            onChange={({ target: { value } }) => setPlaylistName(value)}
            onClick={() => setPlaylistName('')}
            inputProps={{ 'aria-label': 'naked', style: { textAlign: 'center' } }}
            type="text"
          />
        </Tooltip>
        <List>
          {playlist.map(generateListItem(playlist.length, updatePlaylist, classes))}
        </List>
        <Button onClick={onSaveClick} disabled={isLoading || isButtonDisabled} startIcon={isLoading ? null : <SaveIcon />}>
          {isButtonDisabled ? 'Saved' : 'Save Playlist'}
        </Button>
      </CardContent>
    </Card>
  );
};

CustomPlaylistsWidget.defaultProps = {
};

CustomPlaylistsWidget.propTypes = {
  accessToken: PropTypes.string.isRequired,
  userProfile: PropTypes.shape({
    id: PropTypes.string,
  }).isRequired,
  updatePlaylist: PropTypes.shape({
    removeTrack: PropTypes.func.isRequired,
  }).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  playlist: PropTypes.array.isRequired,
};

export default CustomPlaylistsWidget;
