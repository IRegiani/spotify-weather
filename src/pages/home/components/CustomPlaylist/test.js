import React from 'react';
import { shallow } from 'enzyme';

// eslint-disable-next-line no-unused-vars
import CustomPlaylist, { savePlaylist, generateListItem } from './index';

jest.mock('react-router-dom');

describe('CustomPlaylist Widget', () => {
  test('renders correctly', () => {
    const wrapper = shallow(<CustomPlaylist playlist={[]} updatePlaylist={{ removeTrack: jest.fn() }} />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test.skip('renders item correctly', () => {
    const Component = generateListItem()();
    const wrapper = shallow(<Component />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test.todo('save playlist func');
});
