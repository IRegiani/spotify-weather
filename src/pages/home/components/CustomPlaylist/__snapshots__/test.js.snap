// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomPlaylist Widget renders correctly 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-root-1"
  variant="outlined"
>
  <WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(Tooltip))
      arrow={true}
      leaveDelay={500}
      placement="right"
      title="Rename your playlist!"
    >
      <WithStyles(ForwardRef(InputBase))
        inputProps={
          Object {
            "aria-label": "naked",
            "style": Object {
              "textAlign": "center",
            },
          }
        }
        onChange={[Function]}
        onClick={[Function]}
        type="text"
        value="My Playlist"
      />
    </WithStyles(ForwardRef(Tooltip))>
    <WithStyles(ForwardRef(List)) />
    <WithStyles(ForwardRef(Button))
      disabled={false}
      onClick={[Function]}
      startIcon={<Memo(SaveIcon) />}
    >
      Save Playlist
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;
